### PR TITLE
fix: read correct failure status for terminated containers [DET-3435]

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
Previously, for every container termination, we inserted a failure
status, which would result in us terminating the whole gang prematurely.

## Test Plan
Tested this manually.

### CODE REVIEW

File: master/internal/trial.go [hunk 1]
```diff
@@ -741,11 +741,7 @@ func (t *trial) processContainerTerminated(
 		})
 	}
 
-	status := agent.ContainerError(agent.AgentError, errors.New("no error status provided"))
-	if s := msg.ContainerStopped; s.Failure != nil {
-		status = *s
-	}
-	if c == nil || c.IsLeader() || status.Failure != nil {
+	if c == nil || c.IsLeader() || msg.ContainerStopped.Failure != nil {
```
Comment (rb-determined-ai) [inline master/internal/trial.go:744]: Let's put a comment that explains in plain english what we are trying to do with each clause of this if statement.

This whole bug would probably have been avoided if we just explained in a comment what we were trying to do in the first place.

Comment (aaron276h) [reply master/internal/trial.go:746]: Added a comment in here describing the three cases.

Comment (rb-determined-ai) [reply master/internal/trial.go:746]: nit: we are actually terminating the task if the leader exited at all, not just if it was canceled

Also, given the complexity of the state machine, it might be nice to explain under what conditions "the container never started".

Comment (aaron276h) [reply master/internal/trial.go:746]: Done.

Comment (rb-determined-ai) [reply master/internal/trial.go:746]: looks good

Comment (dzhu) [reply master/internal/trial.go:746]: Perhaps the "if the container never started" should also get a tiny bit more explanation, since it's not necessarily immediately clear why the task should be terminated in that case (and I say that as the one who added that condition). I'm thinking "Terminate the task if this container never actually began executing (since the gang will never be able to completely start), if the leader of the gang exited, or if any of the containers exited with a failure."

Comment (aaron276h) [reply master/internal/trial.go:747]: Done.
